### ISSUE TITLE
Disable Cache

### ISSUE THREAD

**Issue by circleupx** [aliostad/CacheCow#244]
Hi, 

I have a question, is there a way to disable CacheCow? 

I would like to disable the cache whenever I am working on my local environment, my current workaround is to force chrome to disable caching. 

Is there a better way to do this? Also, this is an awesome project!

**Comment by aliostad** [comment]
Hi,

Could you please explain more? Do you mean you have a need for configuration based on the environments?

But then again, in your client, you can send `no-cache` header so that the server always send you the fresh resource.

**Comment by circleupx** [comment]
Hi, 

Yes, I guess that is what I am looking for, but just for my local environment. 

I've been doing what you are suggesting, basically, in postman/chrome I sent the appropriate headers to make the server refresh the resources. 

I was just wondering if there was a global disable option within Cachecow because sometimes I forget to add the headers.

Thank you for getting back to me.

**Comment by circleupx** [comment]
One more question, 

Is it possible to set the cache expiration at runtime? Like I would omit the cache expiration from HttpCacheFactory, like so,

```c#
 [HttpCacheFactory(ViewModelType = typeof(Cars))]
```

then at runtime calculate the cache expiration base on the state of the resource.

**Comment by aliostad** [comment]
With the new rich configuration in .NET Core which is environment-aware, I guess we can do something about this now - the project was designed well before any of this.

I will flag it as a feature request and will start looking into it soon and will get back to you. Thanks for the suggestions.

**Comment by circleupx** [comment]
@aliostad 

Do you have any advice on my other question? Is this possible within CacheCow? 

> Is it possible to set the cache expiration at runtime? Like I would omit the cache expiration from HttpCacheFactory then at runtime calculate the cache expiration base on the state of the resource.

I have a REST API that may return multiple resource types base on the client request. 

I would to set the expiration base the resource types within the HTTP response. For example, assume that an API has two resources, car and passenger and I decorate the controller like so,

```C#
public class CarController : Controller
{
    [HttpGet]
    [HttpCacheFactory(300, ViewModelType = typeof(Car))]
    public IActionResult Get(int id)
    {
        ... // implementation
    }
}

public class PassengerController : Controller
{
    [HttpGet]
    [HttpCacheFactory(60, ViewModelType = typeof(Passenger))]
    public IActionResult Get(int id)
    {
        ... // implementation
    }
}
```

The car resource has an expiration of 5 minutes, car information doesn't change that much. Passengers on the other hand is more volatile so I want to keep the cache time low, 1 minute. 

Now, a client makes an HTTP request like so,

https://api.example.com/cars

With the current CacheCow configuration defined above, the server would response with Cache-Control: public, max-age=500.

Cool so far, everything works are expected and the correct cache expiration is set.

Now, assume that the car resource has a one to many relationship to passengers, and through ORM frameworks like EF CORE we can include then on our query should the client ask for resource.

Possible client request may look like this,

https://api.example.com/cars?include=passengers.

With the current CacheCow configuration defined above, the server would response with Cache-Control: public, max-age=500, but in this case I actually want to set the cache expiration to 1 minute because of the inclusion of passengers. 

Thanks!

**Comment by aliostad** [comment]
Thanks for outlining your requirement with a clear example. Can you please open another issue with this particular question so it is also accessible for others?

`ICacheDirectiveProvider` interface and `GetCacheControl` method is responsible for providing the CacheControlHeader. It mainly operates on configuredExpiry but you can override it. Here is the implementation from `DefaultCacheDirectiveProvider`:

``` csharp
#if NET452
        public CacheControlHeaderValue GetCacheControl(HttpActionExecutedContext context, TimeSpan? configuredExpiry)
#else
        public CacheControlHeaderValue GetCacheControl(HttpContext context, TimeSpan? configuredExpiry)
#endif
        {
            switch (configuredExpiry)
            {
                case TimeSpan t when t == TimeSpan.Zero:
                    return new CacheControlHeaderValue() { MaxAge = TimeSpan.Zero, Private = true, MustRevalidate = true };
                case TimeSpan t:
                    return new CacheControlHeaderValue() { MaxAge = t, Public = true, MustRevalidate = true };
                default:
                    return new CacheControlHeaderValue() { NoCache = true, NoStore = true };
            }
        }
    }

```

So you need to create your own implementation which looks at the query string. Context parameter contains all information you need to make a decision.

You need to register your type. If you use the Generic interface means you can override for only when it returns a specific type which targets you method.

I could create an example but I cannot promise when exactly I will be able to.

**Comment by circleupx** [comment]
Thank you, you have been very helpful.

**Comment by aliostad** [comment]
This is now done and will be released in 2.7.0. Have a look at docs here:
https://github.com/aliostad/CacheCow/blob/master/README.md#configuration-options-with-aspnet-core

Cheers
Ali
